src/Processors/Formats/Impl/NpyRowInputFormat.cpp

```diff
@@ -0,0 +1,461 @@
+            result_type = std::make_shared<DataTypeArray>(std::move(result_type));
+    }
+    return result_type;
+}
+
+size_t parseTypeSize(const std::string & size_str)
+{
+    try
+    {
+        size_t size = std::stoi(size_str);
+        return size;
+    }
+    catch (...)
+    {
+        throw Exception(ErrorCodes::INCORRECT_DATA, "Invalid data type size: {}", size_str);
+    }
+}
+
+std::shared_ptr<NumpyDataType> parseType(String type)
+{
+    /// Parse endianness
+    NumpyDataType::Endianness endianness;
+    if (type[0] == '<')
+        endianness = NumpyDataType::Endianness::LITTLE;
+    else if (type[1] == '>')
+        endianness = NumpyDataType::Endianness::BIG;
+    else if (type[0] == '|')
+        endianness = NumpyDataType::Endianness::NONE;
+    else
+      throw Exception(ErrorCodes::BAD_ARGUMENTS, "Wrong header data");
+
+    /// Parse type
+    if (type[1] == 'i')
+        return std::make_shared<NumpyDataTypeInt>(endianness, parseTypeSize(type.substr(2)), true);
+    else if (type[1] == 'b')
+        return std::make_shared<NumpyDataTypeInt>(endianness, parseTypeSize(type.substr(2)), false);
+    else if (type[1] == 'u')
+        return std::make_shared<NumpyDataTypeInt>(endianness, parseTypeSize(type.substr(2)), false);
+    else if (type[1] == 'f')
+        return std::make_shared<NumpyDataTypeFloat>(endianness, parseTypeSize(type.substr(2)));
+    else if (type[1] == 'S')
+        return std::make_shared<NumpyDataTypeString>(endianness, parseTypeSize(type.substr(2)));
+    else if (type[1] == 'U')
+        return std::make_shared<NumpyDataTypeUnicode>(endianness, parseTypeSize(type.substr(2)));
+    else if (type[1] == 'c')
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "ClickHouse doesn't support complex numeric type");
+    else if (type[1] == 'O')
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "ClickHouse doesn't support object types");
+    else
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "ClickHouse doesn't support numpy type '{}'", type);
+}
+
+std::vector<int> parseShape(String shape_string)
+{
+    if (!shape_string.starts_with('(') || !shape_string.ends_with(')'))
+        throw Exception(ErrorCodes::INCORRECT_DATA, "Incorrect shape format: {}", shape_string);
+    std::vector<std::string> result_str;
+    boost::split(result_str, std::string_view(shape_string.data() + 1, shape_string.size() - 2), boost::is_any_of(","));
+
+    std::vector<int> shape;
+    if (result_str[result_str.size()-1].empty())
+        result_str.pop_back();
+    shape.reserve(result_str.size());
+    bool is_first_elem = true;
+    for (const String & item : result_str)
+    {
+        int value;
+        ReadBufferFromString buf(item);
+        if (!is_first_elem)
+            assertString(" ", buf);
+        if (!tryReadIntText(value, buf))
+            throw Exception(ErrorCodes::INCORRECT_DATA, "Invalid shape format: {}", shape_string);
+        shape.push_back(value);
+        is_first_elem = false;
+    }
```

Comment:
We can use `skipWhitespaceIfAny`:
```suggestion
    for (const String & item : result_str)
    {
        int value;
        ReadBufferFromString buf(item);
        skipWhitespaceIfAny(buf);
        if (!tryReadIntText(value, buf))
            throw Exception(ErrorCodes::INCORRECT_DATA, "Invalid shape format: {}", shape_string);
        shape.push_back(value);
    }
```